addon/dom-event-listeners.js

```diff
@@ -0,0 +1,183 @@
+       const el = document.querySelector('.foo');
+       addEventListener(this, el, 'click')
+     }
+   });
+   ```
+
+   @method addEventListener
+   @param { Object } obj the instance to attach the listener for
+   @param { String } selector the DOM selector or element
+   @param { String } _eventName the event name to listen for
+   @param { Function } _callback the callback to run for that event
+   @public
+   */
+export function addEventListener(obj, element, eventName, _callback, options) {
+  assert('Must provide a DOM element when using addEventListener', !!element);
+  assert(
+    'Must provide an element (not a DOM selector) when using addEventListener.',
+    element instanceof Element
+  );
+
+  let callback = run.bind(obj, _callback);
+  let listeners = getEventListeners(obj);
+
+  if (!PASSIVE_SUPPORTED) {
+    options = undefined;
+  }
+
+  element.addEventListener(eventName, callback, options);
+  listeners.push(element, eventName, callback, _callback, options);
+}
+
+/**
+   @param { Object } obj the instance to remove the listener for
+   @param { String } selector the DOM selector or element
+   @param { String } _eventName the event name to listen for
+   @param { Function } callback the callback to run for that event
+   @public
+   */
+export function removeEventListener(
+  obj,
+  element,
+  eventName,
+  callback,
+  options
+) {
+  assert(
+    'Must provide a DOM element when using removeEventListener',
+    !!element
+  );
+  assert(
+    'Must provide an element (not a DOM selector) when using removeEventListener.',
+    element instanceof Element
+  );
+
+  let listeners = getEventListeners(obj);
+
+  if (listeners.length === 0) {
+    return;
+  }
+
+  if (!PASSIVE_SUPPORTED) {
+    options = undefined;
+  }
+
+  // We cannot use Array.findIndex as we cannot rely on babel/polyfill being present
+  for (let i = 0; i < listeners.length; i += LISTENER_ITEM_LENGTH) {
+    if (
+      listeners[i + INDEX.ELEMENT] === element &&
+      listeners[i + INDEX.EVENT_NAME] === eventName &&
+      listeners[i + INDEX.ORIGINAL_CALLBACK] === callback
+    ) {
+      /*
+         * Drop the event listener and remove the listener object
+         */
+      let ownCallback = listeners[i + INDEX.CALLBACK];
+      element.removeEventListener(eventName, ownCallback, options);
+      listeners.splice(i, LISTENER_ITEM_LENGTH);
+      break;
+    }
+  }
+}
+
+function getEventListenersDisposable(eventListeners) {
+  return function() {
+    if (eventListeners !== undefined) {
+      /* Drop non-passive event listeners */
+      for (let i = 0; i < eventListeners.length; i += LISTENER_ITEM_LENGTH) {
+        let element = eventListeners[i + INDEX.ELEMENT];
+        let eventName = eventListeners[i + INDEX.EVENT_NAME];
+        let callback = eventListeners[i + INDEX.CALLBACK];
+        let options = eventListeners[i + INDEX.OPTIONS];
+
+        element.removeEventListener(eventName, callback, options);
```

Comment:
Well they do different things. This explicitly removes all listeners for the object, regardless of type, callback, etc. The `removeEventListener` explicitly filters for the exact match. Seems like a lot of extra work if we were to call `removeEventListener` from in here. Or am I missing something?

Reply:
My concern is that it leaves us in an inconsistent state (e.g. listeners array is not cleared). 

I wasn’t suggesting that this literally call removeEventListener (because we only actually want the second part of that function since we already have all the values from the loop).

Reply:
How is it not cleared? We set it to undefined on line 168.